{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#All combinations\n",
    "\n",
    "import sgkit as sg\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import zarr\n",
    "import xarray as xr\n",
    "import allel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2R analysis\n",
    "\n",
    "ds2 = sg.load_dataset('/Users/christinagkertsou/Desktop/Desktop March/FYP_new_x/dataset/AgamP4_2R.zarr')\n",
    "ds2\n",
    "\n",
    "ds2['sample_cohort'] = xr.DataArray([0] * 2 + [1] * 10 + [2] * 6, dims=\"samples\")\n",
    "ds2['sample_cohort_name'] = xr.DataArray(['ara'] * 2 + ['gam'] * 10 + ['hyb_ctr'] + ['hyb'] * 5, dims=\"samples\")\n",
    "ds2['sample_cohort_name']\n",
    "\n",
    "sample_is_hyb_2R = ds2['sample_cohort_name'] == 'hyb'\n",
    "\n",
    "sample_is_ara_2R = ds2['sample_cohort_name'] == 'ara'\n",
    "\n",
    "sample_is_gam_2R = ds2['sample_cohort_name'] == 'gam'\n",
    "\n",
    "g_2R = allel.GenotypeArray(ds2['call_genotype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>description</th>\n",
       "      <th>Parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_source_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24251444</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>unspecified product</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24251444</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>unspecified product</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24245526</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24245644</td>\n",
       "      <td>24246259</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24247851</td>\n",
       "      <td>24249256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53185</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>47020993</td>\n",
       "      <td>47023594</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP029680-RA-E1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53186</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>CDS</td>\n",
       "      <td>47018304</td>\n",
       "      <td>47020467</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAP029680-PA-CDS2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>AGAP029680-PA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53187</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>CDS</td>\n",
       "      <td>47020993</td>\n",
       "      <td>47021209</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAP029680-PA-CDS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>AGAP029680-PA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53188</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>47017982</td>\n",
       "      <td>47018304</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>utr_AGAP029680-RA_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53189</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>47021209</td>\n",
       "      <td>47023594</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>utr_AGAP029680-RA_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53190 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+------------+---------------------+-----------+-------+\n",
       "| Chromosome   | Source     | Feature             | Start     | +11   |\n",
       "| (category)   | (object)   | (object)            | (int32)   | ...   |\n",
       "|--------------+------------+---------------------+-----------+-------|\n",
       "| AgamP4_2R    | VEuPathDB  | protein_coding_gene | 24245442  | ...   |\n",
       "| AgamP4_2R    | VEuPathDB  | mRNA                | 24245442  | ...   |\n",
       "| AgamP4_2R    | VEuPathDB  | exon                | 24245442  | ...   |\n",
       "| AgamP4_2R    | VEuPathDB  | exon                | 24245644  | ...   |\n",
       "| ...          | ...        | ...                 | ...       | ...   |\n",
       "| AgamP4_2R    | VEuPathDB  | CDS                 | 47018304  | ...   |\n",
       "| AgamP4_2R    | VEuPathDB  | CDS                 | 47020993  | ...   |\n",
       "| AgamP4_2R    | VEuPathDB  | three_prime_UTR     | 47017982  | ...   |\n",
       "| AgamP4_2R    | VEuPathDB  | five_prime_UTR      | 47021209  | ...   |\n",
       "+--------------+------------+---------------------+-----------+-------+\n",
       "Stranded PyRanges object has 53,190 rows and 15 columns from 1 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "11 hidden columns: End, Score, Strand, Frame, ID, description, Parent, gene_id, ... (+ 3 more.)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "# import annotations\n",
    "ann_db_2R = pr.read_gff3(\"/Users/christinagkertsou/Desktop/Finalyearproject/VectorBase-56_AgambiaePEST.gff\")\n",
    "ann_db_2R = ann_db_2R['AgamP4_2R'] \n",
    "ann_db_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ara preparation\n",
    "\n",
    "is_ara_missing_2R = g_2R[:,sample_is_ara_2R].is_missing().all(axis=1)\n",
    "\n",
    "# is_ara_hom_alt = g[:,sample_is_ara].is_hom_alt().sum(axis=1) > 2\n",
    "is_ara_hom_alt_2R = g_2R[:,sample_is_ara_2R].is_hom_alt().all(axis=1)\n",
    "is_ara_hom_ref_2R = g_2R[:,sample_is_ara_2R].is_hom_ref().sum(axis=1) == 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gam preparation\n",
    "\n",
    "is_gam_missing_2R = g_2R[:,sample_is_gam_2R].is_missing().sum(axis=1) > 6\n",
    "is_gam_hom_alt_2R = g_2R[:,sample_is_gam_2R].is_hom_alt().sum(axis=1) > 6\n",
    "is_gam_hom_ref_2R = g_2R[:,sample_is_gam_2R].is_hom_ref().sum(axis=1) > 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "192540"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(is_gam_hom_ref_2R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hyribrids hets\n",
    "\n",
    "hyb_het_2R = g_2R[:,sample_is_hyb_2R].is_het().sum(axis=1) > 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The ara and gam fixed masks\n",
    "\n",
    "mask_ara_fixed_2R = is_ara_hom_alt_2R & (is_gam_missing_2R | is_gam_hom_ref_2R)\n",
    "\n",
    "mask_gam_fixed_2R = is_gam_hom_alt_2R & (is_ara_missing_2R | is_ara_hom_ref_2R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 1: het hybs with mask_gam_fixed_2R\n",
    "\n",
    "mask_hets_2R=hyb_het_2R & mask_gam_fixed_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#positions\n",
    "\n",
    "pos_2R = allel.SortedIndex(ds2['variant_position'])\n",
    "\n",
    "pos_gam_fixed_2R = pos_2R[mask_gam_fixed_2R]\n",
    "\n",
    "pos_hyb_hets_2R = pos_2R[mask_hets_2R]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>description</th>\n",
       "      <th>Parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_source_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>selected_hets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24251444</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>unspecified product</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24251444</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>unspecified product</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24245526</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24245644</td>\n",
       "      <td>24246259</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24247851</td>\n",
       "      <td>24249256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53185</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>47020993</td>\n",
       "      <td>47023594</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP029680-RA-E1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53186</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>CDS</td>\n",
       "      <td>47018304</td>\n",
       "      <td>47020467</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAP029680-PA-CDS2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>AGAP029680-PA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53187</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>CDS</td>\n",
       "      <td>47020993</td>\n",
       "      <td>47021209</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAP029680-PA-CDS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>AGAP029680-PA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53188</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>47017982</td>\n",
       "      <td>47018304</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>utr_AGAP029680-RA_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53189</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>47021209</td>\n",
       "      <td>47023594</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>utr_AGAP029680-RA_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53190 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Source              Feature     Start       End Score  \\\n",
       "0      AgamP4_2R  VEuPathDB  protein_coding_gene  24245442  24251444     .   \n",
       "1      AgamP4_2R  VEuPathDB                 mRNA  24245442  24251444     .   \n",
       "2      AgamP4_2R  VEuPathDB                 exon  24245442  24245526     .   \n",
       "3      AgamP4_2R  VEuPathDB                 exon  24245644  24246259     .   \n",
       "4      AgamP4_2R  VEuPathDB                 exon  24247851  24249256     .   \n",
       "...          ...        ...                  ...       ...       ...   ...   \n",
       "53185  AgamP4_2R  VEuPathDB                 exon  47020993  47023594     .   \n",
       "53186  AgamP4_2R  VEuPathDB                  CDS  47018304  47020467     .   \n",
       "53187  AgamP4_2R  VEuPathDB                  CDS  47020993  47021209     .   \n",
       "53188  AgamP4_2R  VEuPathDB      three_prime_UTR  47017982  47018304     .   \n",
       "53189  AgamP4_2R  VEuPathDB       five_prime_UTR  47021209  47023594     .   \n",
       "\n",
       "      Strand Frame                     ID          description         Parent  \\\n",
       "0          +     .             AGAP002619  unspecified product            NaN   \n",
       "1          +     .          AGAP002619-RA  unspecified product     AGAP002619   \n",
       "2          +     .  exon_AGAP002619-RA-E1                  NaN  AGAP002619-RA   \n",
       "3          +     .  exon_AGAP002619-RA-E2                  NaN  AGAP002619-RA   \n",
       "4          +     .  exon_AGAP002619-RA-E3                  NaN  AGAP002619-RA   \n",
       "...      ...   ...                    ...                  ...            ...   \n",
       "53185      -     .  exon_AGAP029680-RA-E1                  NaN  AGAP029680-RA   \n",
       "53186      -     0     AGAP029680-PA-CDS2                  NaN  AGAP029680-RA   \n",
       "53187      -     0     AGAP029680-PA-CDS1                  NaN  AGAP029680-RA   \n",
       "53188      -     .    utr_AGAP029680-RA_1                  NaN  AGAP029680-RA   \n",
       "53189      -     .    utr_AGAP029680-RA_2                  NaN  AGAP029680-RA   \n",
       "\n",
       "          gene_id protein_source_id Name Note  SNPs  selected_hets  \n",
       "0             NaN               NaN  NaN  NaN     0              0  \n",
       "1             NaN               NaN  NaN  NaN     0              0  \n",
       "2      AGAP002619               NaN  NaN  NaN     0              0  \n",
       "3      AGAP002619               NaN  NaN  NaN     0              0  \n",
       "4      AGAP002619               NaN  NaN  NaN     0              0  \n",
       "...           ...               ...  ...  ...   ...            ...  \n",
       "53185  AGAP029680               NaN  NaN  NaN     0              0  \n",
       "53186  AGAP029680     AGAP029680-PA  NaN  NaN     0              0  \n",
       "53187  AGAP029680     AGAP029680-PA  NaN  NaN     0              0  \n",
       "53188         NaN               NaN  NaN  NaN     0              0  \n",
       "53189         NaN               NaN  NaN  NaN     0              0  \n",
       "\n",
       "[53190 rows x 17 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_2R = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_gam_fixed_2R])\n",
    "pr_hets_2R = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_hyb_hets_2R])\n",
    "\n",
    "count_SNP_df_2R = pr.count_overlaps({'SNPs': pr_gam_fixed_2R, 'selected_hets': pr_hets_2R}, features=ann_db_2R).df\n",
    "count_SNP_df_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>ID</th>\n",
       "      <th>description</th>\n",
       "      <th>Parent</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>protein_source_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>Note</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>selected_hets</th>\n",
       "      <th>selected_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24251444</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>unspecified product</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24251444</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>unspecified product</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24245442</td>\n",
       "      <td>24245526</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24245644</td>\n",
       "      <td>24246259</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>24247851</td>\n",
       "      <td>24249256</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP002619-RA-E3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP002619-RA</td>\n",
       "      <td>AGAP002619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53185</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>exon</td>\n",
       "      <td>47020993</td>\n",
       "      <td>47023594</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>exon_AGAP029680-RA-E1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53186</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>CDS</td>\n",
       "      <td>47018304</td>\n",
       "      <td>47020467</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAP029680-PA-CDS2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>AGAP029680-PA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53187</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>CDS</td>\n",
       "      <td>47020993</td>\n",
       "      <td>47021209</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>AGAP029680-PA-CDS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>AGAP029680</td>\n",
       "      <td>AGAP029680-PA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53188</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>47017982</td>\n",
       "      <td>47018304</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>utr_AGAP029680-RA_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53189</th>\n",
       "      <td>AgamP4_2R</td>\n",
       "      <td>VEuPathDB</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>47021209</td>\n",
       "      <td>47023594</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>utr_AGAP029680-RA_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGAP029680-RA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53190 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Source              Feature     Start       End Score  \\\n",
       "0      AgamP4_2R  VEuPathDB  protein_coding_gene  24245442  24251444     .   \n",
       "1      AgamP4_2R  VEuPathDB                 mRNA  24245442  24251444     .   \n",
       "2      AgamP4_2R  VEuPathDB                 exon  24245442  24245526     .   \n",
       "3      AgamP4_2R  VEuPathDB                 exon  24245644  24246259     .   \n",
       "4      AgamP4_2R  VEuPathDB                 exon  24247851  24249256     .   \n",
       "...          ...        ...                  ...       ...       ...   ...   \n",
       "53185  AgamP4_2R  VEuPathDB                 exon  47020993  47023594     .   \n",
       "53186  AgamP4_2R  VEuPathDB                  CDS  47018304  47020467     .   \n",
       "53187  AgamP4_2R  VEuPathDB                  CDS  47020993  47021209     .   \n",
       "53188  AgamP4_2R  VEuPathDB      three_prime_UTR  47017982  47018304     .   \n",
       "53189  AgamP4_2R  VEuPathDB       five_prime_UTR  47021209  47023594     .   \n",
       "\n",
       "      Strand Frame                     ID          description         Parent  \\\n",
       "0          +     .             AGAP002619  unspecified product            NaN   \n",
       "1          +     .          AGAP002619-RA  unspecified product     AGAP002619   \n",
       "2          +     .  exon_AGAP002619-RA-E1                  NaN  AGAP002619-RA   \n",
       "3          +     .  exon_AGAP002619-RA-E2                  NaN  AGAP002619-RA   \n",
       "4          +     .  exon_AGAP002619-RA-E3                  NaN  AGAP002619-RA   \n",
       "...      ...   ...                    ...                  ...            ...   \n",
       "53185      -     .  exon_AGAP029680-RA-E1                  NaN  AGAP029680-RA   \n",
       "53186      -     0     AGAP029680-PA-CDS2                  NaN  AGAP029680-RA   \n",
       "53187      -     0     AGAP029680-PA-CDS1                  NaN  AGAP029680-RA   \n",
       "53188      -     .    utr_AGAP029680-RA_1                  NaN  AGAP029680-RA   \n",
       "53189      -     .    utr_AGAP029680-RA_2                  NaN  AGAP029680-RA   \n",
       "\n",
       "          gene_id protein_source_id Name Note  SNPs  selected_hets  \\\n",
       "0             NaN               NaN  NaN  NaN     0              0   \n",
       "1             NaN               NaN  NaN  NaN     0              0   \n",
       "2      AGAP002619               NaN  NaN  NaN     0              0   \n",
       "3      AGAP002619               NaN  NaN  NaN     0              0   \n",
       "4      AGAP002619               NaN  NaN  NaN     0              0   \n",
       "...           ...               ...  ...  ...   ...            ...   \n",
       "53185  AGAP029680               NaN  NaN  NaN     0              0   \n",
       "53186  AGAP029680     AGAP029680-PA  NaN  NaN     0              0   \n",
       "53187  AGAP029680     AGAP029680-PA  NaN  NaN     0              0   \n",
       "53188         NaN               NaN  NaN  NaN     0              0   \n",
       "53189         NaN               NaN  NaN  NaN     0              0   \n",
       "\n",
       "       selected_ratio  \n",
       "0                 NaN  \n",
       "1                 NaN  \n",
       "2                 NaN  \n",
       "3                 NaN  \n",
       "4                 NaN  \n",
       "...               ...  \n",
       "53185             NaN  \n",
       "53186             NaN  \n",
       "53187             NaN  \n",
       "53188             NaN  \n",
       "53189             NaN  \n",
       "\n",
       "[53190 rows x 18 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_2R['selected_ratio'] = count_SNP_df_2R['selected_hets'] / count_SNP_df_2R['SNPs']\n",
    "count_SNP_df_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>selected_hets</th>\n",
       "      <th>ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3523</th>\n",
       "      <td>AGAP013395</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103</th>\n",
       "      <td>AGAP002331</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>949</th>\n",
       "      <td>AGAP002153</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3558</th>\n",
       "      <td>AGAP013455</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1237</th>\n",
       "      <td>AGAP002472</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3795</th>\n",
       "      <td>AGAP029806</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3797</th>\n",
       "      <td>AGAP029862</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3801</th>\n",
       "      <td>AGAP029915</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3802</th>\n",
       "      <td>AGAP029916</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3805</th>\n",
       "      <td>AGAP029939</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3806 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_id  SNPs  selected_hets  ratio\n",
       "3523  AGAP013395     1              1    1.0\n",
       "1103  AGAP002331     1              1    1.0\n",
       "949   AGAP002153     1              1    1.0\n",
       "3558  AGAP013455     1              1    1.0\n",
       "1237  AGAP002472     1              1    1.0\n",
       "...          ...   ...            ...    ...\n",
       "3795  AGAP029806     0              0    NaN\n",
       "3797  AGAP029862     0              0    NaN\n",
       "3801  AGAP029915     0              0    NaN\n",
       "3802  AGAP029916     0              0    NaN\n",
       "3805  AGAP029939     0              0    NaN\n",
       "\n",
       "[3806 rows x 4 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_2R = count_SNP_df_2R.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2R = grouped_2R[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2R['ratio'] = grouped_2R['selected_hets'] / grouped_2R['SNPs']\n",
    "grouped_2R.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2R.to_csv(\"Snps_hets_3L.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 2: het hybs with mask_ara_fixed_2R\n",
    "\n",
    "mask_hets_2R_2=hyb_het_2R & mask_ara_fixed_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#positions\n",
    "\n",
    "pos_2R = allel.SortedIndex(ds2['variant_position'])\n",
    "\n",
    "pos_ara_fixed_2R_2 = pos_2R[mask_ara_fixed_2R]\n",
    "\n",
    "pos_hyb_hets_2R_2 = pos_2R[mask_hets_2R_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_2R_2 = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_ara_fixed_2R_2])\n",
    "pr_hets_2R_2 = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_hyb_hets_2R_2])\n",
    "\n",
    "count_SNP_df_2R_2 = pr.count_overlaps({'SNPs': pr_ara_fixed_2R_2, 'selected_hets': pr_hets_2R_2}, features=ann_db_2R).df\n",
    "count_SNP_df_2R_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_SNP_df_2R_2['selected_ratio'] = count_SNP_df_2R_2['selected_hets'] / count_SNP_df_2R_2['SNPs']\n",
    "count_SNP_df_2R_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2R_2 = count_SNP_df_2R_2.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2R_2 = grouped_2R_2[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2R_2 ['ratio'] = grouped_2R_2['selected_hets'] / grouped_2R_2['SNPs']\n",
    "grouped_2R_2.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 3: Pure hybs 1/1 & gam 1/1 & ara -/- | 0/0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pure hyb hom alt\n",
    "hyb_hom_alt_2R=g_2R[:,sample_is_hyb_2R].is_hom_alt().sum(axis=1) > 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1/1 hybs and mask_gam_fixed_2R mask\n",
    "mask_hom_2R_g=hyb_hom_alt_2R & mask_gam_fixed_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#positions\n",
    "\n",
    "pos_2R = allel.SortedIndex(ds2['variant_position'])\n",
    "\n",
    "pos_gam_fixed_2R = pos_2R[mask_gam_fixed_2R]\n",
    "\n",
    "pos_hyb_hom_2R_g = pos_2R[mask_hom_2R_g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_2R = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_gam_fixed_2R])\n",
    "pr_hyb_hom_2R_g = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_hyb_hom_2R_g])\n",
    "\n",
    "count_SNP_df_2R_g = pr.count_overlaps({'SNPs': pr_gam_fixed_2R, 'selected_hets': pr_hyb_hom_2R_g}, features=ann_db_2R).df\n",
    "count_SNP_df_2R_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_2R_g['selected_ratio'] = count_SNP_df_2R_g['selected_hets'] / count_SNP_df_2R_g['SNPs']\n",
    "count_SNP_df_2R_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2R_g = count_SNP_df_2R_g.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2R_g = grouped_2R_g[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2R_g['ratio'] = grouped_2R_g['selected_hets'] / grouped_2R_g['SNPs']\n",
    "grouped_2R_g.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 4: 0/0 | -/- hybs, 1/1 ara, 0/0 | -/- gam\n",
    "hyb_hom_missing_2R=g_2R[:,sample_is_hyb_2R].is_hom_alt().sum(axis=1) > 2\n",
    "hyb_hom_hom_ref_2R=g_2R[:,sample_is_hyb_2R].is_hom_ref().sum(axis=1) > 2\n",
    "\n",
    "hyb_total=hyb_hom_missing_2R | hyb_hom_hom_ref_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#0/0 | -/- hybs and mask_ara_fixed_2R mask\n",
    "mask_hom_ref_2R=hyb_total & mask_ara_fixed_2R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#positions\n",
    "pos_2R = allel.SortedIndex(ds2['variant_position'])\n",
    "\n",
    "pos_ara_fixed_2R= pos_2R[mask_ara_fixed_2R]\n",
    "\n",
    "pos_hyb_hom_ref_2R = pos_2R[mask_hom_ref_2R]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_2R_homs = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_ara_fixed_2R])\n",
    "pr_hyb_hom_ref_2R = pr.from_dict([{'Chromosome': 'AgamP4_2R', 'Start': p, 'End': p} for p in pos_hyb_hom_ref_2R])\n",
    "\n",
    "count_SNP_df_2R_homs_ref = pr.count_overlaps({'SNPs': pr_ara_fixed_2R_homs, 'selected_hets': pr_hyb_hom_ref_2R}, features=ann_db_2R).df\n",
    "count_SNP_df_2R_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "count_SNP_df_2R_homs_ref['selected_ratio'] = count_SNP_df_2R_homs_ref['selected_hets'] / count_SNP_df_2R_homs_ref['SNPs']\n",
    "count_SNP_df_2R_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2R_ara_ref = count_SNP_df_2R_homs_ref.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2R_ara_ref = grouped_2R_ara_ref[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2R_ara_ref['ratio'] = grouped_2R_ara_ref['selected_hets'] / grouped_2R_ara_ref['SNPs']\n",
    "grouped_2R_ara_ref.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2L analysis\n",
    "\n",
    "ds1 = sg.load_dataset('/Users/christinagkertsou/Desktop/Desktop March/FYP_new_x/dataset/AgamP4_2L.zarr')\n",
    "ds1\n",
    "\n",
    "ds1['sample_cohort'] = xr.DataArray([0] * 2 + [1] * 10 + [2] * 6, dims=\"samples\")\n",
    "ds1['sample_cohort_name'] = xr.DataArray(['ara'] * 2 + ['gam'] * 10 + ['hyb_ctr'] + ['hyb'] * 5, dims=\"samples\")\n",
    "ds1['sample_cohort_name']\n",
    "\n",
    "sample_is_hyb_2L = ds1['sample_cohort_name'] == 'hyb'\n",
    "\n",
    "sample_is_ara_2L = ds1['sample_cohort_name'] == 'ara'\n",
    "\n",
    "sample_is_gam_2L = ds1['sample_cohort_name'] == 'gam'\n",
    "\n",
    "g_2L = allel.GenotypeArray(ds1['call_genotype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "# import annotations\n",
    "ann_db_2L = pr.read_gff3(\"/Users/christinagkertsou/Desktop/Finalyearproject/VectorBase-56_AgambiaePEST.gff\")\n",
    "ann_db_2L = ann_db_2L['AgamP4_2L'] \n",
    "ann_db_2L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ara preparation\n",
    "\n",
    "is_ara_missing_2L = g_2L[:,sample_is_ara_2L].is_missing().all(axis=1)\n",
    "\n",
    "# is_ara_hom_alt = g[:,sample_is_ara].is_hom_alt().sum(axis=1) > 2\n",
    "is_ara_hom_alt_2L = g_2L[:,sample_is_ara_2L].is_hom_alt().all(axis=1)\n",
    "is_ara_hom_ref_2L = g_2L[:,sample_is_ara_2L].is_hom_ref().sum(axis=1) == 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gam preparation\n",
    "\n",
    "is_gam_missing_2L = g_2L[:,sample_is_gam_2L].is_missing().sum(axis=1) > 6\n",
    "is_gam_hom_alt_2L = g_2L[:,sample_is_gam_2L].is_hom_alt().sum(axis=1) > 6\n",
    "is_gam_hom_ref_2L = g_2L[:,sample_is_gam_2L].is_hom_ref().sum(axis=1) > 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hyribrids hets\n",
    "\n",
    "hyb_het_2L = g_2L[:,sample_is_hyb_2L].is_het().sum(axis=1) > 2\n",
    "\n",
    "#The ara and gam fixed masks\n",
    "\n",
    "mask_ara_fixed_2L = is_ara_hom_alt_2L & (is_gam_missing_2L | is_gam_hom_ref_2L)\n",
    "\n",
    "mask_gam_fixed_2L = is_gam_hom_alt_2L & (is_ara_missing_2L | is_ara_hom_ref_2L)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 1: het hybs with mask_gam_fixed_2L\n",
    "\n",
    "mask_hets_2L=hyb_het_2L & mask_gam_fixed_2L\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_2L = allel.SortedIndex(ds1['variant_position'])\n",
    "\n",
    "pos_gam_fixed_2L = pos_2L[mask_gam_fixed_2L]\n",
    "\n",
    "pos_hyb_hets_2L = pos_2L[mask_hets_2L]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_2L = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_gam_fixed_2L])\n",
    "pr_hets_2L = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_hyb_hets_2L])\n",
    "\n",
    "count_SNP_df_2L = pr.count_overlaps({'SNPs': pr_gam_fixed_2L, 'selected_hets': pr_hets_2L}, features=ann_db_2L).df\n",
    "count_SNP_df_2L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_2L['selected_ratio'] = count_SNP_df_2L['selected_hets'] / count_SNP_df_2L['SNPs']\n",
    "count_SNP_df_2L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2L = count_SNP_df_2L.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2L = grouped_2L[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2L['ratio'] = grouped_2L['selected_hets'] / grouped_2L['SNPs']\n",
    "grouped_2L.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 2: het hybs with mask_ara_fixed_2L\n",
    "\n",
    "mask_hets_2L_2=hyb_het_2L & mask_ara_fixed_2L\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_2L = allel.SortedIndex(ds1['variant_position'])\n",
    "\n",
    "pos_ara_fixed_2L_2 = pos_2L[mask_ara_fixed_2L]\n",
    "\n",
    "pos_hyb_hets_2L_2 = pos_2L[mask_hets_2L_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_2L_2 = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_ara_fixed_2L_2])\n",
    "pr_hets_2L_2 = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_hyb_hets_2L_2])\n",
    "\n",
    "count_SNP_df_2L_2 = pr.count_overlaps({'SNPs': pr_ara_fixed_2L_2, 'selected_hets': pr_hets_2L_2}, features=ann_db_2L).df\n",
    "count_SNP_df_2L_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_SNP_df_2L_2['selected_ratio'] = count_SNP_df_2L_2['selected_hets'] / count_SNP_df_2L_2['SNPs']\n",
    "count_SNP_df_2L_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2L_2 = count_SNP_df_2L_2.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2L_2 = grouped_2L_2[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2L_2 ['ratio'] = grouped_2L_2['selected_hets'] / grouped_2L_2['SNPs']\n",
    "grouped_2L_2.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 3: Pure hybs 1/1 & gam 1/1 & ara -/- | 0/0\n",
    "\n",
    "#pure hyb hom alt\n",
    "hyb_hom_alt_2L=g_2L[:,sample_is_hyb_2L].is_hom_alt().sum(axis=1) > 2\n",
    "\n",
    "#1/1 hybs and mask_gam_fixed_2L mask\n",
    "mask_hom_2L_g=hyb_hom_alt_2L & mask_gam_fixed_2L\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_2L = allel.SortedIndex(ds1['variant_position'])\n",
    "\n",
    "pos_gam_fixed_2L = pos_2L[mask_gam_fixed_2L]\n",
    "\n",
    "pos_hyb_hom_2L_g = pos_2L[mask_hom_2L_g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_2L = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_gam_fixed_2L])\n",
    "pr_hyb_hom_2L_g = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_hyb_hom_2L_g])\n",
    "\n",
    "count_SNP_df_2L_g = pr.count_overlaps({'SNPs': pr_gam_fixed_2L, 'selected_hets': pr_hyb_hom_2L_g}, features=ann_db_2L).df\n",
    "count_SNP_df_2L_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_2L_g['selected_ratio'] = count_SNP_df_2L_g['selected_hets'] / count_SNP_df_2L_g['SNPs']\n",
    "count_SNP_df_2L_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2L_g = count_SNP_df_2L_g.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2L_g = grouped_2L_g[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2L_g['ratio'] = grouped_2L_g['selected_hets'] / grouped_2L_g['SNPs']\n",
    "grouped_2L_g.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 4: 0/0 | -/- hybs, 1/1 ara, 0/0 | -/- gam\n",
    "hyb_hom_missing_2L=g_2L[:,sample_is_hyb_2L].is_hom_alt().sum(axis=1) > 2\n",
    "hyb_hom_hom_ref_2L=g_2L[:,sample_is_hyb_2L].is_hom_ref().sum(axis=1) > 2\n",
    "\n",
    "hyb_total=hyb_hom_missing_2L | hyb_hom_hom_ref_2L\n",
    "\n",
    "#0/0 | -/- hybs and mask_ara_fixed_2L mask\n",
    "mask_hom_ref_2L=hyb_total & mask_ara_fixed_2L\n",
    "\n",
    "#positions\n",
    "pos_2L = allel.SortedIndex(ds1['variant_position'])\n",
    "\n",
    "pos_ara_fixed_2L= pos_2L[mask_ara_fixed_2L]\n",
    "\n",
    "pos_hyb_hom_ref_2L = pos_2L[mask_hom_ref_2L]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_2L_homs = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_ara_fixed_2L])\n",
    "pr_hyb_hom_ref_2L = pr.from_dict([{'Chromosome': 'AgamP4_2L', 'Start': p, 'End': p} for p in pos_hyb_hom_ref_2L])\n",
    "\n",
    "count_SNP_df_2L_homs_ref = pr.count_overlaps({'SNPs': pr_ara_fixed_2L_homs, 'selected_hets': pr_hyb_hom_ref_2L}, features=ann_db_2L).df\n",
    "count_SNP_df_2L_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "count_SNP_df_2L_homs_ref['selected_ratio'] = count_SNP_df_2L_homs_ref['selected_hets'] / count_SNP_df_2L_homs_ref['SNPs']\n",
    "count_SNP_df_2L_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2L_ara_ref = count_SNP_df_2L_homs_ref.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_2L_ara_ref = grouped_2L_ara_ref[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_2L_ara_ref['ratio'] = grouped_2L_ara_ref['selected_hets'] / grouped_2L_ara_ref['SNPs']\n",
    "grouped_2L_ara_ref.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_2L.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/2L1.csv\")\n",
    "grouped_2L_2.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/2L2.csv\")\n",
    "grouped_2L_g.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/2L3.csv\")\n",
    "grouped_2L_ara_ref.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/2L4.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3L analysis\n",
    "\n",
    "ds3 = sg.load_dataset('/Users/christinagkertsou/Desktop/Desktop March/FYP_new_x/dataset/AgamP4_3L.zarr')\n",
    "ds3\n",
    "\n",
    "ds3['sample_cohort'] = xr.DataArray([0] * 2 + [1] * 10 + [2] * 6, dims=\"samples\")\n",
    "ds3['sample_cohort_name'] = xr.DataArray(['ara'] * 2 + ['gam'] * 10 + ['hyb_ctr'] + ['hyb'] * 5, dims=\"samples\")\n",
    "ds3['sample_cohort_name']\n",
    "\n",
    "sample_is_hyb_3L = ds3['sample_cohort_name'] == 'hyb'\n",
    "\n",
    "sample_is_ara_3L = ds3['sample_cohort_name'] == 'ara'\n",
    "\n",
    "sample_is_gam_3L = ds3['sample_cohort_name'] == 'gam'\n",
    "\n",
    "g_3L = allel.GenotypeArray(ds3['call_genotype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "# import annotations\n",
    "ann_db_3L = pr.read_gff3(\"/Users/christinagkertsou/Desktop/Finalyearproject/VectorBase-56_AgambiaePEST.gff\")\n",
    "ann_db_3L = ann_db_3L['AgamP4_3L'] \n",
    "ann_db_3L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ara preparation\n",
    "\n",
    "is_ara_missing_3L = g_3L[:,sample_is_ara_3L].is_missing().all(axis=1)\n",
    "\n",
    "# is_ara_hom_alt = g[:,sample_is_ara].is_hom_alt().sum(axis=1) > 2\n",
    "is_ara_hom_alt_3L = g_3L[:,sample_is_ara_3L].is_hom_alt().all(axis=1)\n",
    "is_ara_hom_ref_3L = g_3L[:,sample_is_ara_3L].is_hom_ref().sum(axis=1) == 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gam preparation\n",
    "\n",
    "is_gam_missing_3L = g_3L[:,sample_is_gam_3L].is_missing().sum(axis=1) > 6\n",
    "is_gam_hom_alt_3L = g_3L[:,sample_is_gam_3L].is_hom_alt().sum(axis=1) > 6\n",
    "is_gam_hom_ref_3L = g_3L[:,sample_is_gam_3L].is_hom_ref().sum(axis=1) > 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hyribrids hets\n",
    "\n",
    "hyb_het_3L = g_3L[:,sample_is_hyb_3L].is_het().sum(axis=1) > 2\n",
    "\n",
    "#The ara and gam fixed masks\n",
    "\n",
    "mask_ara_fixed_3L = is_ara_hom_alt_3L & (is_gam_missing_3L | is_gam_hom_ref_3L)\n",
    "\n",
    "mask_gam_fixed_3L = is_gam_hom_alt_3L & (is_ara_missing_3L | is_ara_hom_ref_3L)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 1: het hybs with mask_gam_fixed_3L\n",
    "\n",
    "mask_hets_3L=hyb_het_3L & mask_gam_fixed_3L\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_3L = allel.SortedIndex(ds3['variant_position'])\n",
    "\n",
    "pos_gam_fixed_3L = pos_3L[mask_gam_fixed_3L]\n",
    "\n",
    "pos_hyb_hets_3L = pos_3L[mask_hets_3L]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_3L = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_gam_fixed_3L])\n",
    "pr_hets_3L = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_hyb_hets_3L])\n",
    "\n",
    "count_SNP_df_3L = pr.count_overlaps({'SNPs': pr_gam_fixed_3L, 'selected_hets': pr_hets_3L}, features=ann_db_3L).df\n",
    "count_SNP_df_3L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_3L['selected_ratio'] = count_SNP_df_3L['selected_hets'] / count_SNP_df_3L['SNPs']\n",
    "count_SNP_df_3L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3L = count_SNP_df_3L.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3L = grouped_3L[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3L['ratio'] = grouped_3L['selected_hets'] / grouped_3L['SNPs']\n",
    "grouped_3L.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 2: het hybs with mask_ara_fixed_3L\n",
    "\n",
    "mask_hets_3L_2=hyb_het_3L & mask_ara_fixed_3L\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_3L = allel.SortedIndex(ds3['variant_position'])\n",
    "\n",
    "pos_ara_fixed_3L_2 = pos_3L[mask_ara_fixed_3L]\n",
    "\n",
    "pos_hyb_hets_3L_2 = pos_3L[mask_hets_3L_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_3L_2 = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_ara_fixed_3L_2])\n",
    "pr_hets_3L_2 = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_hyb_hets_3L_2])\n",
    "\n",
    "count_SNP_df_3L_2 = pr.count_overlaps({'SNPs': pr_ara_fixed_3L_2, 'selected_hets': pr_hets_3L_2}, features=ann_db_3L).df\n",
    "count_SNP_df_3L_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_SNP_df_3L_2['selected_ratio'] = count_SNP_df_3L_2['selected_hets'] / count_SNP_df_3L_2['SNPs']\n",
    "count_SNP_df_3L_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3L_2 = count_SNP_df_3L_2.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3L_2 = grouped_3L_2[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3L_2 ['ratio'] = grouped_3L_2['selected_hets'] / grouped_3L_2['SNPs']\n",
    "grouped_3L_2.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 3: Pure hybs 1/1 & gam 1/1 & ara -/- | 0/0\n",
    "\n",
    "#pure hyb hom alt\n",
    "hyb_hom_alt_3L=g_3L[:,sample_is_hyb_3L].is_hom_alt().sum(axis=1) > 2\n",
    "\n",
    "#1/1 hybs and mask_gam_fixed_3L mask\n",
    "mask_hom_3L_g=hyb_hom_alt_3L & mask_gam_fixed_3L\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_3L = allel.SortedIndex(ds3['variant_position'])\n",
    "\n",
    "pos_gam_fixed_3L = pos_3L[mask_gam_fixed_3L]\n",
    "\n",
    "pos_hyb_hom_3L_g = pos_3L[mask_hom_3L_g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_3L = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_gam_fixed_3L])\n",
    "pr_hyb_hom_3L_g = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_hyb_hom_3L_g])\n",
    "\n",
    "count_SNP_df_3L_g = pr.count_overlaps({'SNPs': pr_gam_fixed_3L, 'selected_hets': pr_hyb_hom_3L_g}, features=ann_db_3L).df\n",
    "count_SNP_df_3L_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_3L_g['selected_ratio'] = count_SNP_df_3L_g['selected_hets'] / count_SNP_df_3L_g['SNPs']\n",
    "count_SNP_df_3L_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3L_g = count_SNP_df_3L_g.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3L_g = grouped_3L_g[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3L_g['ratio'] = grouped_3L_g['selected_hets'] / grouped_3L_g['SNPs']\n",
    "grouped_3L_g.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 4: 0/0 | -/- hybs, 1/1 ara, 0/0 | -/- gam\n",
    "hyb_hom_missing_3L=g_3L[:,sample_is_hyb_3L].is_hom_alt().sum(axis=1) > 2\n",
    "hyb_hom_hom_ref_3L=g_3L[:,sample_is_hyb_3L].is_hom_ref().sum(axis=1) > 2\n",
    "\n",
    "hyb_total=hyb_hom_missing_3L | hyb_hom_hom_ref_3L\n",
    "\n",
    "#0/0 | -/- hybs and mask_ara_fixed_3L mask\n",
    "mask_hom_ref_3L=hyb_total & mask_ara_fixed_3L\n",
    "\n",
    "#positions\n",
    "pos_3L = allel.SortedIndex(ds3['variant_position'])\n",
    "\n",
    "pos_ara_fixed_3L= pos_3L[mask_ara_fixed_3L]\n",
    "\n",
    "pos_hyb_hom_ref_3L = pos_3L[mask_hom_ref_3L]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_3L_homs = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_ara_fixed_3L])\n",
    "pr_hyb_hom_ref_3L = pr.from_dict([{'Chromosome': 'AgamP4_3L', 'Start': p, 'End': p} for p in pos_hyb_hom_ref_3L])\n",
    "\n",
    "count_SNP_df_3L_homs_ref = pr.count_overlaps({'SNPs': pr_ara_fixed_3L_homs, 'selected_hets': pr_hyb_hom_ref_3L}, features=ann_db_3L).df\n",
    "count_SNP_df_3L_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "count_SNP_df_3L_homs_ref['selected_ratio'] = count_SNP_df_3L_homs_ref['selected_hets'] / count_SNP_df_3L_homs_ref['SNPs']\n",
    "count_SNP_df_3L_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3L_ara_ref = count_SNP_df_3L_homs_ref.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3L_ara_ref = grouped_3L_ara_ref[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3L_ara_ref['ratio'] = grouped_3L_ara_ref['selected_hets'] / grouped_3L_ara_ref['SNPs']\n",
    "grouped_3L_ara_ref.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3L.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3L1.csv\")\n",
    "grouped_3L_2.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3L2.csv\")\n",
    "grouped_3L_g.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3L3.csv\")\n",
    "grouped_3L_ara_ref.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3L4.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3R analysis\n",
    "ds4 = sg.load_dataset('/Users/christinagkertsou/Desktop/Desktop March/FYP_new_x/dataset/AgamP4_3R.zarr')\n",
    "ds4\n",
    "\n",
    "ds4['sample_cohort'] = xr.DataArray([0] * 2 + [1] * 10 + [2] * 6, dims=\"samples\")\n",
    "ds4['sample_cohort_name'] = xr.DataArray(['ara'] * 2 + ['gam'] * 10 + ['hyb_ctr'] + ['hyb'] * 5, dims=\"samples\")\n",
    "ds4['sample_cohort_name']\n",
    "\n",
    "sample_is_hyb_3R = ds4['sample_cohort_name'] == 'hyb'\n",
    "\n",
    "sample_is_ara_3R = ds4['sample_cohort_name'] == 'ara'\n",
    "\n",
    "sample_is_gam_3R = ds4['sample_cohort_name'] == 'gam'\n",
    "\n",
    "g_3R = allel.GenotypeArray(ds4['call_genotype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "# import annotations\n",
    "ann_db_3R = pr.read_gff3(\"/Users/christinagkertsou/Desktop/Finalyearproject/VectorBase-56_AgambiaePEST.gff\")\n",
    "ann_db_3R = ann_db_3R['AgamP4_3R'] \n",
    "ann_db_3R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ara preparation\n",
    "\n",
    "is_ara_missing_3R = g_3R[:,sample_is_ara_3R].is_missing().all(axis=1)\n",
    "\n",
    "# is_ara_hom_alt = g[:,sample_is_ara].is_hom_alt().sum(axis=1) > 2\n",
    "is_ara_hom_alt_3R = g_3R[:,sample_is_ara_3R].is_hom_alt().all(axis=1)\n",
    "is_ara_hom_ref_3R = g_3R[:,sample_is_ara_3R].is_hom_ref().sum(axis=1) == 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gam preparation\n",
    "\n",
    "is_gam_missing_3R = g_3R[:,sample_is_gam_3R].is_missing().sum(axis=1) > 6\n",
    "is_gam_hom_alt_3R = g_3R[:,sample_is_gam_3R].is_hom_alt().sum(axis=1) > 6\n",
    "is_gam_hom_ref_3R = g_3R[:,sample_is_gam_3R].is_hom_ref().sum(axis=1) > 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hyribrids hets\n",
    "\n",
    "hyb_het_3R = g_3R[:,sample_is_hyb_3R].is_het().sum(axis=1) > 2\n",
    "\n",
    "#The ara and gam fixed masks\n",
    "\n",
    "mask_ara_fixed_3R = is_ara_hom_alt_3R & (is_gam_missing_3R | is_gam_hom_ref_3R)\n",
    "\n",
    "mask_gam_fixed_3R = is_gam_hom_alt_3R & (is_ara_missing_3R | is_ara_hom_ref_3R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 1: het hybs with mask_gam_fixed_3R\n",
    "\n",
    "mask_hets_3R=hyb_het_3R & mask_gam_fixed_3R\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_3R = allel.SortedIndex(ds4['variant_position'])\n",
    "\n",
    "pos_gam_fixed_3R = pos_3R[mask_gam_fixed_3R]\n",
    "\n",
    "pos_hyb_hets_3R = pos_3R[mask_hets_3R]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_3R = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_gam_fixed_3R])\n",
    "pr_hets_3R = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_hyb_hets_3R])\n",
    "\n",
    "count_SNP_df_3R = pr.count_overlaps({'SNPs': pr_gam_fixed_3R, 'selected_hets': pr_hets_3R}, features=ann_db_3R).df\n",
    "count_SNP_df_3R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_3R['selected_ratio'] = count_SNP_df_3R['selected_hets'] / count_SNP_df_3R['SNPs']\n",
    "count_SNP_df_3R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3R = count_SNP_df_3R.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3R = grouped_3R[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3R['ratio'] = grouped_3R['selected_hets'] / grouped_3R['SNPs']\n",
    "grouped_3R.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 2: het hybs with mask_ara_fixed_3R\n",
    "\n",
    "mask_hets_3R_2=hyb_het_3R & mask_ara_fixed_3R\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_3R = allel.SortedIndex(ds4['variant_position'])\n",
    "\n",
    "pos_ara_fixed_3R_2 = pos_3R[mask_ara_fixed_3R]\n",
    "\n",
    "pos_hyb_hets_3R_2 = pos_3R[mask_hets_3R_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_3R_2 = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_ara_fixed_3R_2])\n",
    "pr_hets_3R_2 = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_hyb_hets_3R_2])\n",
    "\n",
    "count_SNP_df_3R_2 = pr.count_overlaps({'SNPs': pr_ara_fixed_3R_2, 'selected_hets': pr_hets_3R_2}, features=ann_db_3R).df\n",
    "count_SNP_df_3R_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_SNP_df_3R_2['selected_ratio'] = count_SNP_df_3R_2['selected_hets'] / count_SNP_df_3R_2['SNPs']\n",
    "count_SNP_df_3R_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3R_2 = count_SNP_df_3R_2.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3R_2 = grouped_3R_2[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3R_2 ['ratio'] = grouped_3R_2['selected_hets'] / grouped_3R_2['SNPs']\n",
    "grouped_3R_2.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 3: Pure hybs 1/1 & gam 1/1 & ara -/- | 0/0\n",
    "\n",
    "#pure hyb hom alt\n",
    "hyb_hom_alt_3R=g_3R[:,sample_is_hyb_3R].is_hom_alt().sum(axis=1) > 2\n",
    "\n",
    "#1/1 hybs and mask_gam_fixed_3R mask\n",
    "mask_hom_3R_g=hyb_hom_alt_3R & mask_gam_fixed_3R\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_3R = allel.SortedIndex(ds4['variant_position'])\n",
    "\n",
    "pos_gam_fixed_3R = pos_3R[mask_gam_fixed_3R]\n",
    "\n",
    "pos_hyb_hom_3R_g = pos_3R[mask_hom_3R_g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_3R = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_gam_fixed_3R])\n",
    "pr_hyb_hom_3R_g = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_hyb_hom_3R_g])\n",
    "\n",
    "count_SNP_df_3R_g = pr.count_overlaps({'SNPs': pr_gam_fixed_3R, 'selected_hets': pr_hyb_hom_3R_g}, features=ann_db_3R).df\n",
    "count_SNP_df_3R_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_3R_g['selected_ratio'] = count_SNP_df_3R_g['selected_hets'] / count_SNP_df_3R_g['SNPs']\n",
    "count_SNP_df_3R_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3R_g = count_SNP_df_3R_g.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3R_g = grouped_3R_g[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3R_g['ratio'] = grouped_3R_g['selected_hets'] / grouped_3R_g['SNPs']\n",
    "grouped_3R_g.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 4: 0/0 | -/- hybs, 1/1 ara, 0/0 | -/- gam\n",
    "hyb_hom_missing_3R=g_3R[:,sample_is_hyb_3R].is_hom_alt().sum(axis=1) > 2\n",
    "hyb_hom_hom_ref_3R=g_3R[:,sample_is_hyb_3R].is_hom_ref().sum(axis=1) > 2\n",
    "\n",
    "hyb_total=hyb_hom_missing_3R | hyb_hom_hom_ref_3R\n",
    "\n",
    "#0/0 | -/- hybs and mask_ara_fixed_3R mask\n",
    "mask_hom_ref_3R=hyb_total & mask_ara_fixed_3R\n",
    "\n",
    "#positions\n",
    "pos_3R = allel.SortedIndex(ds4['variant_position'])\n",
    "\n",
    "pos_ara_fixed_3R= pos_3R[mask_ara_fixed_3R]\n",
    "\n",
    "pos_hyb_hom_ref_3R = pos_3R[mask_hom_ref_3R]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_3R_homs = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_ara_fixed_3R])\n",
    "pr_hyb_hom_ref_3R = pr.from_dict([{'Chromosome': 'AgamP4_3R', 'Start': p, 'End': p} for p in pos_hyb_hom_ref_3R])\n",
    "\n",
    "count_SNP_df_3R_homs_ref = pr.count_overlaps({'SNPs': pr_ara_fixed_3R_homs, 'selected_hets': pr_hyb_hom_ref_3R}, features=ann_db_3R).df\n",
    "count_SNP_df_3R_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "count_SNP_df_3R_homs_ref['selected_ratio'] = count_SNP_df_3R_homs_ref['selected_hets'] / count_SNP_df_3R_homs_ref['SNPs']\n",
    "count_SNP_df_3R_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3R_ara_ref = count_SNP_df_3R_homs_ref.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_3R_ara_ref = grouped_3R_ara_ref[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_3R_ara_ref['ratio'] = grouped_3R_ara_ref['selected_hets'] / grouped_3R_ara_ref['SNPs']\n",
    "grouped_3R_ara_ref.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_3R.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3R1.csv\")\n",
    "grouped_3R_2.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3R2.csv\")\n",
    "grouped_3R_g.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3R3.csv\")\n",
    "grouped_3R_ara_ref.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/3R4.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#X analysis\n",
    "\n",
    "ds5 = sg.load_dataset('/Users/christinagkertsou/Desktop/Desktop March/FYP_new_x/dataset/AgamP4_X.zarr')\n",
    "ds5\n",
    "\n",
    "ds5['sample_cohort'] = xr.DataArray([0] * 2 + [1] * 10 + [2] * 6, dims=\"samples\")\n",
    "ds5['sample_cohort_name'] = xr.DataArray(['ara'] * 2 + ['gam'] * 10 + ['hyb_ctr'] + ['hyb'] * 5, dims=\"samples\")\n",
    "ds5['sample_cohort_name']\n",
    "\n",
    "sample_is_hyb_X = ds5['sample_cohort_name'] == 'hyb'\n",
    "\n",
    "sample_is_ara_X = ds5['sample_cohort_name'] == 'ara'\n",
    "\n",
    "sample_is_gam_X = ds5['sample_cohort_name'] == 'gam'\n",
    "\n",
    "g_X = allel.GenotypeArray(ds5['call_genotype'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "# import annotations\n",
    "ann_db_X = pr.read_gff3(\"/Users/christinagkertsou/Desktop/Finalyearproject/VectorBase-56_AgambiaePEST.gff\")\n",
    "ann_db_X = ann_db_X['AgamP4_X'] \n",
    "ann_db_X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ara preparation\n",
    "\n",
    "is_ara_missing_X = g_X[:,sample_is_ara_X].is_missing().all(axis=1)\n",
    "\n",
    "# is_ara_hom_alt = g[:,sample_is_ara].is_hom_alt().sum(axis=1) > 2\n",
    "is_ara_hom_alt_X = g_X[:,sample_is_ara_X].is_hom_alt().all(axis=1)\n",
    "is_ara_hom_ref_X = g_X[:,sample_is_ara_X].is_hom_ref().sum(axis=1) == 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gam preparation\n",
    "\n",
    "is_gam_missing_X = g_X[:,sample_is_gam_X].is_missing().sum(axis=1) > 6\n",
    "is_gam_hom_alt_X = g_X[:,sample_is_gam_X].is_hom_alt().sum(axis=1) > 6\n",
    "is_gam_hom_ref_X = g_X[:,sample_is_gam_X].is_hom_ref().sum(axis=1) > 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hyribrids hets\n",
    "\n",
    "hyb_het_X = g_X[:,sample_is_hyb_X].is_het().sum(axis=1) > 2\n",
    "\n",
    "#The ara and gam fixed masks\n",
    "\n",
    "mask_ara_fixed_X = is_ara_hom_alt_X & (is_gam_missing_X | is_gam_hom_ref_X)\n",
    "\n",
    "mask_gam_fixed_X = is_gam_hom_alt_X & (is_ara_missing_X | is_ara_hom_ref_X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 1: het hybs with mask_gam_fixed_X\n",
    "\n",
    "mask_hets_X=hyb_het_X & mask_gam_fixed_X\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_X = allel.SortedIndex(ds5['variant_position'])\n",
    "\n",
    "pos_gam_fixed_X = pos_X[mask_gam_fixed_X]\n",
    "\n",
    "pos_hyb_hets_X = pos_X[mask_hets_X]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_X = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_gam_fixed_X])\n",
    "pr_hets_X = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_hyb_hets_X])\n",
    "\n",
    "count_SNP_df_X = pr.count_overlaps({'SNPs': pr_gam_fixed_X, 'selected_hets': pr_hets_X}, features=ann_db_X).df\n",
    "count_SNP_df_X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_X['selected_ratio'] = count_SNP_df_X['selected_hets'] / count_SNP_df_X['SNPs']\n",
    "count_SNP_df_X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_X = count_SNP_df_X.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_X = grouped_X[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_X['ratio'] = grouped_X['selected_hets'] / grouped_X['SNPs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 2: het hybs with mask_ara_fixed_X\n",
    "\n",
    "mask_hets_X_2=hyb_het_X & mask_ara_fixed_X\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_X = allel.SortedIndex(ds5['variant_position'])\n",
    "\n",
    "pos_ara_fixed_X_2 = pos_X[mask_ara_fixed_X]\n",
    "\n",
    "pos_hyb_hets_X_2 = pos_X[mask_hets_X_2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_X_2 = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_ara_fixed_X_2])\n",
    "pr_hets_X_2 = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_hyb_hets_X_2])\n",
    "\n",
    "count_SNP_df_X_2 = pr.count_overlaps({'SNPs': pr_ara_fixed_X_2, 'selected_hets': pr_hets_X_2}, features=ann_db_X).df\n",
    "count_SNP_df_X_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_SNP_df_X_2['selected_ratio'] = count_SNP_df_X_2['selected_hets'] / count_SNP_df_X_2['SNPs']\n",
    "count_SNP_df_X_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_X_2 = count_SNP_df_X_2.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_X_2 = grouped_X_2[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_X_2 ['ratio'] = grouped_X_2['selected_hets'] / grouped_X_2['SNPs']\n",
    "grouped_X_2.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 3: Pure hybs 1/1 & gam 1/1 & ara -/- | 0/0\n",
    "\n",
    "#pure hyb hom alt\n",
    "hyb_hom_alt_X=g_X[:,sample_is_hyb_X].is_hom_alt().sum(axis=1) > 2\n",
    "\n",
    "#1/1 hybs and mask_gam_fixed_X mask\n",
    "mask_hom_X_g=hyb_hom_alt_X & mask_gam_fixed_X\n",
    "\n",
    "#positions\n",
    "\n",
    "pos_X = allel.SortedIndex(ds5['variant_position'])\n",
    "\n",
    "pos_gam_fixed_X = pos_X[mask_gam_fixed_X]\n",
    "\n",
    "pos_hyb_hom_X_g = pos_X[mask_hom_X_g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_gam_fixed_X = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_gam_fixed_X])\n",
    "pr_hyb_hom_X_g = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_hyb_hom_X_g])\n",
    "\n",
    "count_SNP_df_X_g = pr.count_overlaps({'SNPs': pr_gam_fixed_X, 'selected_hets': pr_hyb_hom_X_g}, features=ann_db_X).df\n",
    "count_SNP_df_X_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "\n",
    "count_SNP_df_X_g['selected_ratio'] = count_SNP_df_X_g['selected_hets'] / count_SNP_df_X_g['SNPs']\n",
    "count_SNP_df_X_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_X_g = count_SNP_df_X_g.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_X_g = grouped_X_g[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_X_g['ratio'] = grouped_X_g['selected_hets'] / grouped_X_g['SNPs']\n",
    "grouped_X_g.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Scenario 4: 0/0 | -/- hybs, 1/1 ara, 0/0 | -/- gam\n",
    "hyb_hom_missing_X=g_X[:,sample_is_hyb_X].is_hom_alt().sum(axis=1) > 2\n",
    "hyb_hom_hom_ref_X=g_X[:,sample_is_hyb_X].is_hom_ref().sum(axis=1) > 2\n",
    "\n",
    "hyb_total=hyb_hom_missing_X | hyb_hom_hom_ref_X\n",
    "\n",
    "#0/0 | -/- hybs and mask_ara_fixed_X mask\n",
    "mask_hom_ref_X=hyb_total & mask_ara_fixed_X\n",
    "\n",
    "#positions\n",
    "pos_X = allel.SortedIndex(ds5['variant_position'])\n",
    "\n",
    "pos_ara_fixed_X= pos_X[mask_ara_fixed_X]\n",
    "\n",
    "pos_hyb_hom_ref_X = pos_X[mask_hom_ref_X]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "pr_ara_fixed_X_homs = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_ara_fixed_X])\n",
    "pr_hyb_hom_ref_X = pr.from_dict([{'Chromosome': 'AgamP4_X', 'Start': p, 'End': p} for p in pos_hyb_hom_ref_X])\n",
    "\n",
    "count_SNP_df_X_homs_ref = pr.count_overlaps({'SNPs': pr_ara_fixed_X_homs, 'selected_hets': pr_hyb_hom_ref_X}, features=ann_db_X).df\n",
    "count_SNP_df_X_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios\n",
    "count_SNP_df_X_homs_ref['selected_ratio'] = count_SNP_df_X_homs_ref['selected_hets'] / count_SNP_df_X_homs_ref['SNPs']\n",
    "count_SNP_df_X_homs_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_X_ara_ref = count_SNP_df_X_homs_ref.query(\"Feature == 'exon'\").groupby('gene_id').aggregate(sum).reset_index()\n",
    "grouped_X_ara_ref = grouped_X_ara_ref[['gene_id', 'SNPs', 'selected_hets']]\n",
    "grouped_X_ara_ref['ratio'] = grouped_X_ara_ref['selected_hets'] / grouped_X_ara_ref['SNPs']\n",
    "grouped_X_ara_ref.sort_values('ratio', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_X_2.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/X2.csv\")\n",
    "grouped_X_g.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/X3.csv\")\n",
    "grouped_X_ara_ref.to_csv(\"/Users/christinagkertsou/Desktop/ola_gambiae/X4.csv\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c1fea38084ca1a29ef4291c83b38a58ab7adc3ad74c330bf0268bcc7479f7557"
  },
  "kernelspec": {
   "display_name": "Python 3.8.12 ('hybrids')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
